deputy.go

```diff
@@ -0,0 +1,182 @@
+	// simply returns the error from Cmd.Run()
+	DefaultErrs ErrorHandling = iota
+
+	// FromStderr tells Deputy to convert the stderr output of a command into
+	// the text of an error, if the command exits with an error.
+	FromStderr
+
+	// FromStdout tells Deputy to convert the stdout output of a command into
+	// the text of an error, if the command exits with an error.
+	FromStdout
+)
+
+// Deputy is a type that runs Commands with advanced options not available from
+// os/exec.  See the comments on field values for details.
+type Deputy struct {
+	// Timeout represents the longest time the command will be allowed to run
+	// before being killed.
+	Timeout time.Duration
+	// Errors describes how errors should be handled.
+	Errors ErrorHandling
+	// StdoutLog takes a function that will receive lines written to stdout from
+	// the command.
+	StdoutLog func(string)
+	// StdoutLog takes a function that will receive lines written to stderr from
+	// the command.
+	StderrLog func(string)
+
+	stderrPipe io.ReadCloser
+	stdoutPipe io.ReadCloser
+}
+
+// Run starts the specified command and waits for it to complete.  Its behavior
+// conforms to the Options passed to it at construction time.
+func (d Deputy) Run(cmd *exec.Cmd) error {
+	if d.StderrLog != nil {
+		var err error
+		d.stderrPipe, err = cmd.StderrPipe()
+		if err != nil {
+			return err
+		}
+	}
+	if d.StdoutLog != nil {
+		var err error
+		d.stdoutPipe, err = cmd.StdoutPipe()
+		if err != nil {
+			return err
+		}
+	}
+
+	errsrc := &bytes.Buffer{}
+	if d.Errors == FromStderr {
+		if cmd.Stderr == nil {
+			cmd.Stderr = errsrc
+		} else {
+			cmd.Stderr = io.MultiWriter(cmd.Stderr, errsrc)
+		}
+	}
+	if d.Errors == FromStdout {
+		if cmd.Stdout == nil {
+			cmd.Stdout = errsrc
+		} else {
+			cmd.Stdout = io.MultiWriter(cmd.Stdout, errsrc)
+		}
+	}
+
+	err := d.run(cmd)
+
+	if d.Errors == DefaultErrs {
+		return err
+	}
+
+	if err != nil && errsrc.Len() > 0 {
+		return fmt.Errorf("%s: %s", err, strings.TrimSpace(errsrc.String()))
+	}
+	return err
+}
+
+func (d Deputy) run(cmd *exec.Cmd) error {
+	errs := make(chan error)
+	if err := d.start(cmd, errs); err != nil {
+		return err
+	}
+	if d.Timeout == 0 {
+		return d.wait(cmd, errs)
+	}
+
+	done := make(chan error)
+
+	var err error
+	go func() {
+		err = d.wait(cmd, errs)
+		close(done)
+	}()
+
+	select {
+	case <-time.After(d.Timeout):
+		// this may fail, but there's not much we can do about it
+		_ = cmd.Process.Kill()
+		return timeoutErr{cmd.Path}
+	case <-done:
+		return err
+	}
+}
+
+func (d Deputy) start(cmd *exec.Cmd, errs chan<- error) error {
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	if d.stdoutPipe != nil {
+		go pipe(d.StdoutLog, d.stdoutPipe, errs)
+	}
+	if d.stderrPipe != nil {
+		go pipe(d.StderrLog, d.stderrPipe, errs)
+	}
+	return nil
+}
+
+func (d Deputy) wait(cmd *exec.Cmd, errs <-chan error) error {
+	var err1, err2 error
+	if d.stdoutPipe != nil {
```

Comment:
// Note that it's important that we wait for the pipes
// to be closed before calling cmd.Wait otherwise
// Wait can close the pipes before we have read
// all their data.

?